java/query/src/main/java/sleeper/query/lambda/SqsLeafPartitionQueryLambda.java

```diff
@@ -0,0 +1,91 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import sleeper.configuration.jars.ObjectFactoryException;
+import sleeper.configuration.properties.instance.InstanceProperties;
+import sleeper.configuration.properties.table.TablePropertiesProvider;
+import sleeper.query.tracker.DynamoDBQueryTracker;
+
+import static sleeper.configuration.properties.instance.CdkDefinedInstanceProperty.CONFIG_BUCKET;
+
+/**
+ * A lambda that is triggered when a serialised leaf partition query arrives on an SQS queue. A processor executes the request using a
+ * {@link sleeper.query.recordretrieval.LeafPartitionQueryExecutor} and publishes the results to either SQS or S3 based on the configuration
+ * of the query. The processor contains a cache that includes mappings from partitions to files in those partitions. This is reused by
+ * subsequent calls to the lambda if the AWS runtime chooses to reuse the instance.
+ */
+@SuppressWarnings("unused")
+public class SqsLeafPartitionQueryLambda implements RequestHandler<SQSEvent, Void> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SqsLeafPartitionQueryLambda.class);
+
+    private final AmazonSQS sqsClient;
+    private final AmazonS3 s3Client;
+    private final AmazonDynamoDB dynamoClient;
+    private long lastUpdateTime;
+    private InstanceProperties instanceProperties;
+    private SqsLeafPartitionQueryProcessor processor;
+    private TablePropertiesProvider tablePropertiesProvider;
+    private QueryMessageHandler messageHandler;
+
+    public SqsLeafPartitionQueryLambda() throws ObjectFactoryException {
+        this(AmazonS3ClientBuilder.defaultClient(), AmazonSQSClientBuilder.defaultClient(),
+                AmazonDynamoDBClientBuilder.defaultClient(), System.getenv(CONFIG_BUCKET.toEnvironmentVariable()));
+    }
+
+    public SqsLeafPartitionQueryLambda(AmazonS3 s3Client, AmazonSQS sqsClient, AmazonDynamoDB dynamoClient, String configBucket) throws ObjectFactoryException {
+        this.s3Client = s3Client;
+        this.sqsClient = sqsClient;
+        this.dynamoClient = dynamoClient;
+        instanceProperties = loadInstanceProperties(s3Client, configBucket);
+        tablePropertiesProvider = new TablePropertiesProvider(instanceProperties, s3Client, dynamoClient);
+        messageHandler = new QueryMessageHandler(tablePropertiesProvider, new DynamoDBQueryTracker(instanceProperties, dynamoClient));
+        processor = SqsLeafPartitionQueryProcessor.builder()
+            .sqsClient(sqsClient).s3Client(s3Client).dynamoClient(dynamoClient)
+            .instanceProperties(instanceProperties).tablePropertiesProvider(tablePropertiesProvider)
+            .build();
+    }
+
+    @Override
+    public Void handleRequest(SQSEvent event, Context context) {
+        for (SQSEvent.SQSMessage message : event.getRecords()) {
+            LOGGER.info("Received message with body {}", message.getBody());
+            event.getRecords().stream()
```

Comment:
This is a nested loop. Instead of processing each message once, this is processing each message as many times as the number of messages that were received. You could either remove the for loop or the streaming.

Reply:
This is copied from the existing SqsQueryProcessorLambda class. Is this an existing issue with the other class?

Reply:
The batch size is set to 1 in QueryStack against the trigger for the lambda, so it's not a bug for the system as a whole. It only shows up in the tests since that's the only place multiple messages are given to the lambda at once. It still seems worth fixing though, since the tests are asserting incorrect behaviour.

Reply:
It seems worth fixing this for both lambdas. There's no test where this causes a problem for the non-leaf query lambda though. We'd need a test that specifically sends multiple queries at the same time. Up to you whether you want to add that.